services/blockstorage/adapter/filesystem/construct_index_test.go

```diff
@@ -52,6 +53,71 @@ func TestConstructIndexFromReader(t *testing.T) {
 
 }
 
+func newBlockFileReadStream(t *testing.T, ctrlRand *rand.ControlledRand, numBlocks int32, codec *codec) (io.Reader, chan int) {
+	blocksQueue := builders.RandomizedBlockChain(numBlocks, ctrlRand)
+
+	pr, pw := io.Pipe()
+
+	done := make(chan int)
+	go func() {
+		for _, blockPair := range blocksQueue {
+			_, err := codec.encode(blockPair, pw)
+			require.NoError(t, err, "expected codec to successfully encode a block pair")
+		}
+		_ = pw.Close()
+		close(done)
+	}()
+
+	return pr, done
+}
+
+func TestBuildIndexSucceedsIndexingFromReader(t *testing.T) {
+	with.Logging(t, func(harness *with.LoggingHarness) {
+		const numBlocks = 17
+
+		ctrlRand := rand.NewControlledRand(t)
+		codec := newCodec(1024 * 1024)
+		r, done := newBlockFileReadStream(t, ctrlRand, numBlocks, codec)
+
+		bhIndex, err := buildIndex(r, 0, harness.Logger, codec)
+
+		require.NoError(t, err, "expected buildIndex to succeed")
+		require.Equal(t, bhIndex.topBlockHeight, primitives.BlockHeight(numBlocks), "expected block height to match the encoded block count")
+
+		<-done
+	})
+}
+
+func TestBuildIndexHandlesSmallBufferedReader(t *testing.T) {
+	with.Logging(t, func(harness *with.LoggingHarness) {
+		const numBlocks = 17
+
+		ctrlRand := rand.NewControlledRand(t)
+		codec := newCodec(1024 * 1024)
+		r, done := newBlockFileReadStream(t, ctrlRand, numBlocks, codec)
+
+		rBuffered := bufio.NewReaderSize(r, 1)
+		bhIndex, err := buildIndex(rBuffered, 0, harness.Logger, codec)
```

Comment:
I think it does test the right behaviour (`buildIndex` handling partial reads) - the buffered reader's `Read` method will only call the underlying reader's `Read` once, resulting in an actual buffer size of the smaller buffer. But since this is very implementation specific I agree it's best not to rely on it.

Reply:
We can increase the number of blocks in the test chain to say, 70, instead of 7, which will practically assure this.

Reply:
not sure how this will help it.